aten/src/ATen/native/FunctionOfAMatrixUtils.cpp

```diff
@@ -0,0 +1,105 @@
+#include <ATen/native/FunctionOfAMatrixUtils.h>
+
+namespace at { namespace native {
+
+DEFINE_DISPATCH(_compute_linear_combination_stub);
+
+// If `coefficients` is a [m, n] Tensor and
+// `input` is a [n, ...] Tensor, then the output
+// `output` is going to be a [m, ...] Tensor such that
+// for i in range(m):
+//    for j in range(n):
+//        output[i, ...] += coefficients[i, j] * input[j, ...]
```

Comment:
I still get the feeling that this can be implemented directly (in fused form) using our matrix multiply primitives...

Reply:
Yes, via broadcasting, but it will require m x n x ... memory, and we can do it in O((m + n + ....)

Reply:
Have a look how I test the correctness in `test_matrix_exp_utils`. This fused version gives about 80-100% boost compared to the brodcasted version plus is more memory efficient.

Reply:
```
import torch

# input: [n, ...]
# coefficients: [m, n]
def linear_combination(input, coefficients):
    # input_t: [..., n, 1]
    input_t = torch.movedim(input, 0, -1).unsqueeze(-1)
    # r: [..., m, 1]
    r = torch.matmul(coefficients, input_t)
    # [m, ...]
    return torch.movedim(r.squeeze(-1), -1, 0)

N = 3
M = 2
B1 = 5
B2 = 7

input = torch.randn(N, B1, B2)
coefficients = torch.randn(M, N)
print(linear_combination(input, coefficients))
```

Since this gets to use the blas matrix multiply should be faster too.

Reply:
Yes, but the running time is then the product of the dimensions, not the sum of dimensions.

Reply:
cc @zou3519 I had to compute the movedims and unsqueezes manually, but this seems like a great example of code that could be conveniently computed with vmap

Reply:
So, we get O(m * n^2 * ...) versus O(m * n * ....)

Reply:
Docs for `movedim` are missing.

Reply:
(for the record, I didn't see any of nikitaved's comments as I was working on these, damn you GitHub!)

Reply:
After looking over your tests, I'm still a bit confused. In your test, the reference implementation is

```
 res2 = (x.unsqueeze(0) * coeffs.view(*coeff_shape, *([1] * len(data_shape)))).sum(1)
```

This looks way worse than the matmul pseudocode I posted above...

Reply:
It is better in terms of time complexity, but worse in terms of memory complexity, yes.

Reply:
The complexity of individual matrix multiply in batched matrix multiply is m*n, so times number of batches we'd also get O(m*n*...), but depending on how this case is implemented internally in `matmul` we might be materializing `coefficients` `...` times

Reply:
I can modify the test if you think it is worth it.

Reply:
@ngimel, ah, correct, it becomes a matrix-vector product, by bad.

Reply:
Also, I get a significant boost from using `__restrict__` modifier here.

Reply:
I guess, I'm curious to know if the matmul version loses performance or not. If it *does* lose performance, then this suggests (A) there is a case in matmul that is less performant than it should be, and (B) maybe we should rejigger matmul to make use of your linear combination implementation in this specific case (definitely not in this PR).

I can believe that `__restrict__` helps here, but in general writing optimized matrix-vector products is serious business and if the operation we requested maps to an actual BLAS primitive, I still generally expect the BLAS version to beat anything you would have written here.

Reply:
@ngimel suggests in https://github.com/pytorch/pytorch/pull/40853#discussion_r448975081 that matmul might be doing something stupid, which I could definitely believe

Reply:
Ok, let me benchmark then.

Reply:
All right, time-wise the BLAS routine is definitely faster, but do you know whether it is memory efficient in terms of how it handles `coefficients` in the batched setup?

Reply:
Luckily, for this case (3+D x 2D) we have a special optimization that  reduces memory consumption:
https://github.com/pytorch/pytorch/blob/27e8dc78ca4a87984e7def69213cabdb39fb751f/aten/src/ATen/native/LinearAlgebra.cpp#L582-L600
It still calls `.contiguous()` on the 3+D tensor, which would most likely make a copy (after `movedim`), but it's still better than I expected :-)

Reply:
All right, then this optimization does not take place in the script by @ezyang , because it is actually the first argument which is a 2D matrix, so `matmul` will invoke `bmm`, or am I wrong? I will update the script and test against the transposed result then.

Reply:
This case is handled in the next conditional https://github.com/pytorch/pytorch/blob/27e8dc78ca4a87984e7def69213cabdb39fb751f/aten/src/ATen/native/LinearAlgebra.cpp#L600-L622

Reply:
Oh, that is great then! :)

Reply:
But they could have swapped the arguments based on their inner dimensions to reduce the number of cases and make the code easier to read.

Reply:
Here are some preliminary results. Coefficients are of size `batch x batch`. We test very small coefficient sizes, as sizes beyond `5 x 5` are not used in this PR.

<details>
<summary> CPU </summary>

```
CPU BLAS mm
                |  64x64  |  256x256  |  1024x1024  |  4096x4096  |  16384x16384
24 threads: --------------------------------------------------------------------
      batch: 2  |   200   |    600    |     1000    |     50000   |     600000
      batch: 3  |   210   |    610    |     1000    |     63000   |     850000
      batch: 4  |   300   |    620    |     1500    |     76000   |    1100000
      batch: 5  |   210   |    500    |     2000    |     97000   |    1300000
      batch: 6  |   210   |    410    |     3300    |    110000   |    2000000
      batch: 7  |   230   |    500    |     4700    |    140000   |    2000000
Times are in microseconds (us).

CPU _compute_linear_combination
                |  64x64  |  256x256  |  1024x1024  |  4096x4096  |  16384x16384
24 threads: --------------------------------------------------------------------
      batch: 2  |    16   |    100    |      310    |     30000   |     430000
      batch: 3  |    26   |    130    |      400    |     51000   |     700000
      batch: 4  |    39   |    160    |     1000    |     78000   |    1100000
      batch: 5  |    54   |    190    |     3000    |    100000   |    1500000
      batch: 6  |    75   |    260    |     5000    |    130000   |    2000000
      batch: 7  |   104   |    200    |     6000    |    170000   |    2000000
Times are in microseconds (us).
```

</details>

<details>
<summary>GPU</summary>

```
GPU BLAS mm
                |  64x64  |  256x256  |  1024x1024  |  4096x4096  |  16384x16384
24 threads: --------------------------------------------------------------------
      batch: 2  |    64   |     42    |     187     |     3090    |      49770
      batch: 3  |    42   |     45    |     207     |     3590    |      79800
      batch: 4  |    42   |     42    |     223     |     4100    |      94780
      batch: 5  |    42   |     42    |     242     |     4590    |     137000
      batch: 6  |    42   |     42    |     300     |     5600    |     186000
      batch: 7  |    43   |     42    |     275     |     5560    |     215000
Times are in microseconds (us).

GPU _compute_linear_combination
                |  64x64  |  256x256  |  1024x1024  |  4096x4096  |  16384x16384
24 threads: --------------------------------------------------------------------
      batch: 2  |    16   |     16    |      78     |     1152    |      18370
      batch: 3  |    16   |     16    |     135     |     2084    |      33280
      batch: 4  |    16   |     16    |     211     |     3300    |      52660
      batch: 5  |    16   |     16    |     302     |     4770    |      76520
      batch: 6  |    16   |     16    |     412     |     6508    |     104000
      batch: 7  |    16   |     16    |     533     |     8462    |     135400
Times are in microseconds (us).

```

</details>

`_compute_linear_combination` should be more memory efficient, this is why it is written in the first place.
If we are to compute the highest degree approximation for an input of size `k`, `_compute_linear_combination` will use only `5k` memory. `matmul` will call `contiguous` on the inputs, and `contiguous` on the output. Besides, the result of this linear combination has to be contiguous, so, after changing the axes, I will need to call `contiguous` again. So, we can will need at least `10k` memory. So, I believe, having `_compute_linear_combination` is justifiable exactly because of the memory reasons, the reason it was written for.

Reply:
Besides, the time complexity of this operation compared to matrix-matrix products is still relatively small.

Reply:
Something weird happens with GitHub, other users' comments show up later in-between :)

Reply:
Yeah. So given these results, I'm fine to land this PR with this part of the algorithm. However, it would be pretty cool if this linear combination algorithm could be incorporated into the matmul algorithm selection so *everyone* gets the memory savings in this case. This might not be too easy to do, as we can see from your performance results that which is better depends on the size of the kernel in question. Or perhaps, there is some way to fiddle with the batched mm blas calls so we don't have to expand. That would be the most ideal situation...

Reply:
> It still calls .contiguous() on the 3+D tensor, which would most likely make a copy (after movedim), but it's still better than I expected :-)

BTW, I just assumed that we were trying to implement this operation with the batch dimensions at the end. To tell the truth, that's pretty weird, and I wasn't sure if you could rearrange the call site so that you don't have to do this movedim.